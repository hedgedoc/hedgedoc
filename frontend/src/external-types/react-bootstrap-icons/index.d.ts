/*
 * SPDX-FileCopyrightText: 2023 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */

declare module 'react-bootstrap-icons/dist/icons/*' {
  import type { Icon } from 'react-bootstrap-icons'
  const ReactComponent: Icon
  export default ReactComponent
}
