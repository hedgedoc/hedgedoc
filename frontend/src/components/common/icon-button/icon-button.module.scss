/*
 * SPDX-FileCopyrightText: 2021 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */

.btn-icon {

  &.with-border {
    .text-part {
      border-left: 1px solid rgba(0, 0, 0, 0.2);
      padding-left: var(--bs-btn-padding-x);
    }
  }

  gap: var(--bs-btn-padding-x);

  .icon-part {
    display: flex;
    align-items: center;
    justify-content: center;
  }

  .text-part {
    display: flex;
    align-items: center;
  }
}

