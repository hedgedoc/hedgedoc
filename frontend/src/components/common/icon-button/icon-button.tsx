/*
 * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import { concatCssClasses } from '../../../utils/concat-css-classes'
import type { PropsWithDataTestId } from '../../../utils/test-id'
import { testId } from '../../../utils/test-id'
import { UiIcon } from '../icons/ui-icon'
import { ShowIf } from '../show-if/show-if'
import styles from './icon-button.module.scss'
import React, { useMemo } from 'react'
import type { ButtonProps } from 'react-bootstrap'
import { Button } from 'react-bootstrap'
import type { Icon } from 'react-bootstrap-icons'

export interface IconButtonProps extends ButtonProps, PropsWithDataTestId {
  icon: Icon
  border?: boolean
  iconSize?: number | string
}

/**
 * A generic {@link Button button} with an icon in it.
 *
 * @param icon Which icon should be used
 * @param children The children that will be added as the content of the button.
 * @param iconFixedWidth If the icon should be of fixed width.
 * @param border Should the button have a border.
 * @param className Additional class names added to the button.
 * @param iconSize Size of the icon
 * @param props Additional props for the button.
 */
export const IconButton: React.FC<IconButtonProps> = ({
  icon,
  children,
  border = false,
  className,
  iconSize,
  ...props
}) => {
  const finalClassName = useMemo(
    () =>
      concatCssClasses(styles['btn-icon'], 'd-inline-flex align-items-stretch', className, {
        [styles['with-border']]: border
      }),
    [border, className]
  )

  return (
    <Button {...props} className={finalClassName} {...testId('icon-button')}>
      <span className={`${styles['icon-part']}`}>
        <UiIcon size={iconSize} icon={icon} className={'icon'} />
      </span>
      <ShowIf condition={!!children}>
        <span className={`${styles['text-part']}`}>{children}</span>
      </ShowIf>
    </Button>
  )
}
