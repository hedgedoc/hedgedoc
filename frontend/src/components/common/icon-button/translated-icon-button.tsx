/*
 * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import type { IconButtonProps } from './icon-button'
import { IconButton } from './icon-button'
import React from 'react'
import { Trans } from 'react-i18next'

export interface TranslatedIconButtonProps extends IconButtonProps {
  i18nKey: string
}

/**
 * Renders an {@link IconButton icon button} with a translation inside.
 *
 * @param i18nKey The key for the translated string.
 * @param props Additional props directly given to the {@link IconButton}.
 */
export const TranslatedIconButton: React.FC<TranslatedIconButtonProps> = ({ i18nKey, ...props }) => {
  return (
    <IconButton {...props}>
      <Trans i18nKey={i18nKey} />
    </IconButton>
  )
}
