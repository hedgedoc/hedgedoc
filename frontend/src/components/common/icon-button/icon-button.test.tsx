/*
 * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import { IconButton } from './icon-button'
import { fireEvent, render, screen } from '@testing-library/react'
import { Heart as IconHeart } from 'react-bootstrap-icons'

describe('IconButton', () => {
  it('renders heart icon', () => {
    const view = render(<IconButton icon={IconHeart}>test</IconButton>)
    expect(view.container).toMatchSnapshot()
  })
  it('renders with border', () => {
    const view = render(
      <IconButton icon={IconHeart} border={true}>
        test with border
      </IconButton>
    )
    expect(view.container).toMatchSnapshot()
  })
  it('renders with additional className', () => {
    const view = render(
      <IconButton icon={IconHeart} className={'testClass'}>
        test with additional className
      </IconButton>
    )
    expect(view.container).toMatchSnapshot()
  })
  it('correctly uses the onClick callback', async () => {
    const onClick = jest.fn()
    const view = render(
      <IconButton icon={IconHeart} onClick={onClick}>
        test with onClick
      </IconButton>
    )
    expect(view.container).toMatchSnapshot()
    const iconButton = await screen.findByTestId('icon-button')
    fireEvent(
      iconButton,
      new MouseEvent('click', {
        bubbles: true,
        cancelable: true
      })
    )
    expect(onClick).toHaveBeenCalled()
  })
})
