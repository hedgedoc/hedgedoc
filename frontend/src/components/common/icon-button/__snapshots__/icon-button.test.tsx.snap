// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IconButton correctly uses the onClick callback 1`] = `
<div>
  <button
    class="btn-icon p-0 d-inline-flex align-items-stretch   btn btn-primary"
    data-testid="icon-button"
    type="button"
  >
    <span
      class="icon-part d-flex align-items-center"
    >
      BootstrapIconMock_Heart
    </span>
    <span
      class="text-part d-flex align-items-center"
    >
      test with onClick
    </span>
  </button>
</div>
`;

exports[`IconButton renders heart icon 1`] = `
<div>
  <button
    class="btn-icon p-0 d-inline-flex align-items-stretch   btn btn-primary"
    data-testid="icon-button"
    type="button"
  >
    <span
      class="icon-part d-flex align-items-center"
    >
      BootstrapIconMock_Heart
    </span>
    <span
      class="text-part d-flex align-items-center"
    >
      test
    </span>
  </button>
</div>
`;

exports[`IconButton renders with additional className 1`] = `
<div>
  <button
    class="btn-icon p-0 d-inline-flex align-items-stretch  testClass btn btn-primary"
    data-testid="icon-button"
    type="button"
  >
    <span
      class="icon-part d-flex align-items-center"
    >
      BootstrapIconMock_Heart
    </span>
    <span
      class="text-part d-flex align-items-center"
    >
      test with additional className
    </span>
  </button>
</div>
`;

exports[`IconButton renders with border 1`] = `
<div>
  <button
    class="btn-icon p-0 d-inline-flex align-items-stretch with-border  btn btn-primary"
    data-testid="icon-button"
    type="button"
  >
    <span
      class="icon-part d-flex align-items-center"
    >
      BootstrapIconMock_Heart
    </span>
    <span
      class="text-part d-flex align-items-center"
    >
      test with border
    </span>
  </button>
</div>
`;
